### ISSUE TITLE
Introduce a way to detect repeatable events

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#59]
In https://github.com/WordPoints/hooks-api/issues/53#issuecomment-167155946 it has come to my attention that we really have two different types of hook events: repeatable and non-repeatable. This distinction is important for several reasons:
- Non-repeatable events should only be allowed to hit once per-arg, unless the previous hit was reversed. #53 
- We currently only want to use the periods extension for repeatable events. #10 

An example of a repeatable event would be User Visit: a user can visit multiple times and each time be awarded points.

An example of a non-repeatable event would be Publish Post: a user can toggle a post's status from `publish` to `draft` and back as many times as they want, but they'll only be awarded points once (for that post).

Unfortunately, there really is no good way to detect whether an event is repeatable or not in the API. As discussed in the comment linked to above, this really shouldn't be tied to reversibility. What I suggested there was introducing an interface that could be implemented by the classes for repeatable events.

**Comment by JDGrimes** [comment]
## What is a repetition?

One thing that we need to define is what exactly constitutes a repeat fire for a non-repeatable event. 

First, it isn't that the event is only allowed to fire once, but that it is only allowed to _hit_ once. And if that hit is reversed/spammed, a new hit is allowed to occur.

Also, it isn't that the event is only allowed to hit once, period; it is that it is allowed to hit only once in relation to each entity. That is, it is only allowed to hit once per each value of the event arg.

So using the example above, a user is allowed to be awarded once per post that they publish by the Publish Post event.

The only thing we need to consider is events with multiple args. Since there really is no concept of "main arg" in the API, do we consider a fire a repeat only if all of the args are the same? I guess that makes sense. But the question then is, what about when args are added or removed from the event? The fires will then have a different "signature", even if the remaining args are the same. This is an issue that came up when considering handling reversals in the periods API (something we decided not to do at the moment). The proposed solution was to store the values of all of the (top level) event args and then only compare the args in the current fire that were also in the past fire. Then we'd consider it a repeat only if all of them are the same.

However, I'm not entirely convinced that we should be taking the values of all of the args into account. We could have a "stateful" arg registered that isn't really really important to consider. For example, we don't really need to take the current user or something like that into account when deciding whether a Post Publish event is being repeated. But like I said above, we don't really have a good way to determine which arg(s) are the "main" ones for the event. So I guess if we want to have this API that's another thing that we'll need to do.

**Comment by JDGrimes** [comment]
So I guess what we could do is differentiate the repeatable events from the rest by the fact that they didn't have any "main" args (I guess we could call them "signature" args). Then we wouldn't need the interface.

**Comment by JDGrimes** [comment]
## Are there non-signature args?

Do we currently have any args that wouldn't be considered signature args?
- `current:user` arg for the User Visit event.
- `current:site` arg on multisite.

The `current:site` arg would be the only non-signature arg registered for non-repeatable events at present. However, I'm not even sure that it should be considered a non-signature arg. For standard reactions on multisite the current site will always be the same whenever that reaction fires. So since the arg value will never change, the point is really moot (since it would never cause the signature of the fire to change). But for network reactions it is a different story. The signature would change from one site to the next. But is that expected behavior? I guess that would depend on how network reactions should behave. What is considered a repeat of a network event?

## What is a network event repeat?

That's a difficult thing to define, because most events occur in context of a specific site. Each post belongs to a specific site, for example. I guess the only entity that we currently deal with that exists across the network would be users. User data is stored in a single global table, although whether a user is a member of a particular site is determined on a per-site basis. The important thing is that the user ID refers the to the same ting across the entire network. So the user registration event would be a network-context event (unless we make it behave differently on multisite so that it is about becoming a member of a site and not registering on the network). Which means each user registers only once, and there's really no way that that can repeat.

Taking another tack, think about it this way: if we had a user that published a post with the ID 5 on one site, and then published another post with the ID of 5 on another site, would we consider that a repeat fire of the event? Of course not! So it is fine that the event signature is different because the `current:site` arg has different values.

I wonder whether that would be true for every event though. What if there was a type of entity where we wanted to count the same IDs on different sites as repeats? I guess that would usually make sense only when the entity was network-wide itself (like users).

What did we do historically in a case like that? Actually, we didn't have to worry about it. Repeats were handled on a per hook type basis. Some types of points hook, like user register, are tied to actions that will only ever fire once for the user anyway, so detecting repeats isn't necessary. For the post and comment related events, what we did was to run a points logs query or store metadata, respectively. The points logs query would run for logs only for the current site, so that's correct. And of course comment metadata is stored for each particular comment on whatever site that it is on. So again, there was no danger of flagging repeats from other sites.

The only reason that we're running into this issue now is that we're trying to introduce a generic API for this. It will work fine with the events relating to posts or other entities that are site specific. It will also work fine with events relating to users and other global entities, as long as those events only ever fire once anyway. The question is what will happen in the future when we try to introduce a hypothetical event that is tied to a global entity like users but can occur multiple times on different sites.

I guess one example would be becoming a member of a site. That relates to the user, a global entity, but it can happen on each site on the network. And we'd want it to fire for each site on the network. In other words, it would be OK if the event signature was different on different sites. But are there cases where we wouldn't want that to be so? Are there cases where the event would be global to the network? And if so, then wouldn't we just forgo registering the `current:site` arg for that particular event?

The question is, how could we ever have an event that was global to the network (like registration) but also happened in the context of different sites multiple times? That doesn't even make sense. If it can happen multiple times on different sites, it doesn't sound like it is in the global context of the whole network. So an event like that could never really exist.

To answer the question we asked above then, yes, we would just skip registering the `current:site` arg for an event that was in the context of the entire network. Although I suppose it might make sense to register it for the user register event, if we wanted to award the admin of the site on which registration originally took place. But on the other hand, that is an event that only occurs once anyway, so it isn't important to our discussion regarding repeats.

So what defines a network reaction repeat? If the event args include the `current:site` arg, repeats would be calculated per site. If it does not, they would be calculated in context of the entire network.

Actually, the `current:site` reference is not the defining thing; we might also say that any event which had a post as one of its args, or any other non–network-wide entity, should have repeats calculated per site. The only problem is that we have no way at present of detecting whether an entity is global to the network or not, so it is easier for us to just define this in terms of the `current:site` or `site` entity specifically. Otherwise we'd need to introduce some means of determining whether each entity was global or not (though perhaps that would be useful in other parts of the API as well).

The main takeaway is that given the current state of the event args and entity APIs, the `current:site` arg is actually an important part of the signature when determining whether network-wide reaction fires are repeats or not.

As for the possibility of network-context events for which we'd want to register the `current:site` arg (and hence would want it to be a non-signature arg because we're in network context): Most network-wide events take place in the network admin or other network context when we're not actually in the context of a specific site. I think it would actually make sense to disable the `current:site` arg when in network-mode.

In other words, the `current:site` arg isn't actually a non-signature arg, but an important (or at least non-detrimental) part of the signature.

**Comment by JDGrimes** [comment]
Although the above discussion was very productive, it didn't actually answer the question of whether there are any non-signature args. We did decide that the `current:site` arg is actually needed as a part of the signature at the moment, although this may be handled differently in the future. That leaves only the `current:user` arg. However, the `current:user` arg is currently only used for the User Visit event, which is a repeatable event. Mostly we're interested in non-signature args for non-repeatable events.

## What defines a signature arg?

To help us decide whether such a thing could or should exist (in the future sometime, since we don't really have any now), we should clarify exactly what makes an arg a signature arg.

A signature arg is an arg that is integral to the event. The event could not occur at all without that entity being involved. For example, every publish post event fire is inextricably tied to a particular post. It wouldn't even make sense to talk about the event fire outside of the context of that post. The post is by no means a mere "bystander" that just happens to be affected by the event, it is the defining entity of the event. None of the other specifics matter. It doesn't matter how many times that post's status is toggled by different people, or how many different UI's it is edited through—all of those events are really just the same thing, since they have the same result: the post is published.

We might say then, that if we introduced a `current:user` arg for the post publish event, that such an arg would be non-signature, because it isn't a defining aspect of the event. Such an arg would be useful, because we could then award points to editors who triage and publish posts. But we wouldn't want to award each editor who toggled the post status from publish to draft and back; the post itself can only be published once, so no matter whether the `current:user` arg was different, points should still be awarded only once.

What this tells us is that we might indeed have a use for non-signature args in the future beyond just the repeatable events. We might ask, however, whether this would include anything other than the `current:user` arg.

In answer to that, let's consider a hypothetical `current:post\post` arg for an event. That's something that we know would be useful for the User Visit event, but what about non-repeatable events?

OK, let's think about this: if an entity is integral to the event, it will either be related to one of the args, or it will be one of the args. If it isn't integral to the event, and is related to one of the other args, it doesn't need to be registered as an arg itself. So the only time we'd need to register a non-signature arg for an event would be if it wasn't related to any of the other args. If can't be related to the other event args in the database, that means that it is a purely stateful arg. Therefore, we might conclude that it is only stateful args that would ever be non-signature args. (Note that the converse is not true however, some stateful args can be signature args, such as `current:site`.)

But what about when the args aren't related in the database but the event is logged in some way and they can be related through that? They could only be related in the context of that event, so we wouldn't be able to relate them to each other as we normally do in the UI. There isn't a relationship between the entities, as between a post and its author; there is a relationship between the entities and the event. The relationships that we currently deal with are entity to entity, though I guess we could probably do the same thing in the context of the event.

To better explain what I mean, consider an example. User A edits User B. When that happens, User B is the main entity of the event, represented by the `user` arg. User A is just a bystander, if you will, represented by the `current:user` arg. But this is a repeatable event, is it not? So I guess the point is moot. But what I was thinking is that we normally wouldn't consider User A as integral to the event, it would be a non-signature arg, and wouldn't be related to User B in the database. So of course it is a stateful arg. But then what if we had some code that logged each user edit event (outside of WordPoints's hooks API itself) and kept track of which user edited who. Now when the event occurs we could actually relate these two users, through the event. But it doesn't make sense to show the editor as `User » Editor`, we should just display it as `Editor`. So hacking together a special kind of relationship here would seem to produce something other than what we actually want.

I guess the truth is that for every event the args are related to each other through the event, but its only a run-time relationship. And whether it is logged to the DB or not doesn't change that they should both be event args, not one arg related to another entity. Then we'd also need a special relationship for an entity that only is registered for certain events and special handling in the UI. Yikes!

So the above example is actually not what I thought, and just proves that every arg that isn't integral to the event or related to another arg will be a stateful arg. And whether the event gets saved in the database somehow somewhere does nothing to change that.

### Conclusion

Every arg that isn't integral to the event or related to another arg will be a stateful arg. Therefore all non-signature args will also be stateful args.

Non-signature args and stateful args are really the same thing, we've just been approaching it from two different perspectives.

**Comment by JDGrimes** [comment]
What I'm trying to figure out here is whether we need to mark repeatable events in some way or whether we can determine whether an event is repeatable based on its args. If we need to mark some args for even non-repeatable events as signature args and others not, we could just identify a repeatable event by the fact that it would have no signature args.

So what we're trying to figure out is whether we need to mark the args for non-repeatable events or not. Is it possible that there will be non-signature args for non-repeatable events? If not it would be simpler to just mark the repeatable events as repeatable, since they would be the only ones with non-signature args anyway.

We've tried to answer that question in the above two comments, and here's what we've come up with:
- We don't currently register an non-signature args for non-repeatable events.
- All non-signature args, whether for repeatable or non-repeatable events, are stateful args.
- It is possible that we will need stateful/non-signature args for non-repeatable events in the future. (For example, to award the editor who published a post.)

This means that we can't define a non-repeatable event simply by a lack of signature/stateful args. So if we're going to define repeatable events based on their args, we need to ask whether the converse is true:

## Can repeatable events have non-stateful args?

To answer this question, let's consider what the nature of such an arg would be. Above, we said this:

> Every arg that isn't integral to the event or related to another arg will be a stateful arg. Therefore all non-signature args will also be stateful args.

So, in converse, this would imply that if it isn't a stateful arg, then it must be integral to the event. By "integral", it was meant that the event was specific to a certain entity (the post that was published by the Publish Post event, for example). For the event to be repeatable though, it means that it can happen multiple times without being reversed. That implies that the event doesn't actually change the state of any particular entities. A post can't go from being unpublished to being published twice, without going from published to unpublished first. But if the event doesn't modify the entity in any way, then isn't it a bystander? Why should that entity be a part of the signature?

That isn't to say that we'd never want to limit the User Visit hook to fire only once per post, for example. But that this is something that isn't an intrinsic quality of the event itself. That is, we could accomplish this through the limits extension, but it is an arbitrary added constraint. We wouldn't always want to consider the ID of the post being viewed when determining whether the event was a repeat.

So what makes a repeatable event repeatable? It is that it doesn't modify the state of any of the entities involved. I suppose we might say that this is even something separate from reversibility. An event could be reversible and yet still repeatable. For example, we might log something like user visits, and allow them to be marked as spam. This would allow specific hits to be reversed. But the event would still be repeatable without reversals occurring.

I guess we'd say it this way: a repeatable event is allowed to occur repeatedly without all prior hits being reversed.

All we're trying to figure out is whether this tells us anything about its args, and more importantly, whether the converse is also true. Would we have a repeatable event with signature args? Put another way: Would it ever make sense to have an event be repeatable when the state of one of the entities changes when the event occurs? Actually, I guess that would usually be impossible, since the event couldn't occur a second time without first being reversed. Which means it couldn't be a repeatable event in the first place.

This excludes the possibility of having a repeatable event with args whose state is toggled by the event.

But this still leaves a few loopholes that we need to close:
- Can events alter the state of an entity in other ways beside "toggling" such that the event can be reversed but can also be repeated without being reversed?
- Are all signature args' states changed when the event occurs?

### Are signature args always toggled?

The last question first. The significance of it is that we've concluded that no repeatable event can have an arg whose state is toggled by that event. If all signature args must have their state toggled when an event occurs, then we could conclude it was impossible for a repeatable event to have a signature arg. And since signature args and stateful args are the same, we could thus identify a repeatable event by a lack of stateful/signature args.

So, does being a signature arg imply that the state of the arg is toggled when the event occurs? What makes an arg a signature arg?

Above we concluded that some signature args could be stateful args. So we could ask whether a stateful arg is ever going to have its state toggled by an event. If not, that would answer our question; signature args wouldn't necessarily have their states toggled.

The only stateful signature arg that we have considered is the `current:site` arg. And that is just a hack for proper handling of network reaction fires. So really, we don't have any real cases of stateful signature args. The conclusion is that there may really be no such thing as stateful signature args.

What would such an arg be like? If would have to be integral to the event (and thus a signature arg) and yet be unrelated to any of the other args in the event. It seems like then it would probably be one of the "main" args of the event. In other words, we'd consider it part of the event signature (since it's a signature arg, after all). But then, what makes it a stateful arg? Because its status isn't toggled? But then what makes it integral?

I think perhaps the best way to define a signature arg is by the fact that its status is toggled. However, before we decide that, there's one more thing we need to look at.

### Non-toggling entity alterations

What about when a repeatable event modifies an entity in some way, but the modification isn't a "toggle" of some status. Might the entity still be considered a signature arg of the event?

We could have an event that fires when a user is edited. The edited user is modified, but not necessarily their status. The edited user wouldn't be considered a signature arg of the event though, because it is possible to edit the user multiple times. We could of course use the limits extension to change that for a particular reaction, but it isn't a limit that needs to be intrinsic to the event, as pointed out above.

The whole premise here is untenable. How can the event be repeatable but still have this signature arg? What we're looking for is an event that is relative to a particular arg, but isn't repeatable for that arg. It is only repeatable for other values. But then it isn't really a repeatable event is it. It is then a non-repeatable event. 

In other words:
- We know that repeatable events cannot be tied to args whose status is toggled by the event.
- Changes that occur to an entity during an even other than toggling the status are allowed to happen.
- But such changes are by nature repeatable, so the args for such entities wouldn't be signature args.

And most importantly: an event isn't repeatable if it cannot be repeated relative to _all_ of its args. That is to say, a non-repeatable event can have non-signature args, but since signature args imply non-repeatability, the converse is not also true: a repeatable event cannot have signature args.

To answer our original question, "Can repeatable events have non-stateful args?": No.

**Comment by JDGrimes** [comment]
This means that we could, in theory, use the args of an event to determine whether it was repeatable, given one more assumption: non-repeatable events will always have at least one signature arg. This is easily proven by the fact that non-repeatable events must toggle the state of some entity, otherwise they wouldn't be non-repeatable. They will therefore always have at least one signature arg.

## Context-specific entities

There is one more issue that needs to be considered when defining event repeatability: some entities are not global, but context-specific.

What is meant by the entity being "local" to a particular context is that entities of that type can exist within the scope of the reaction, such that two such entities with the same ID may exist in separate sub-contexts.

We discussed this above with network-wide reactions and posts. Posts are entities that exist in the context of a specific site, not the network as a whole. This means that each site may have a post with the ID 2, for example, but because these are different posts, each with its own status, even when an event is non-repeatable, the event isn't restricted to occurring only once per post ID, but once per post per site on the network.

Stated generally, a fire of an event is only considered a repeat if it has already occurred for an entity with the ID of the signature arg _in the context in which that entity exists_.

In other words, the repeat must be calculated relative to that particular entity, not just to its ID.

The question, then, is how we do should do this. Should we try to introduce a full-fledged "contexts" API? Or should we just allow the entity class to supply a totally unique ID that can be used across all contexts?

Or, are we really going to have other contexts beside the site context that we run into on multisite? If not, we could build the API specifically for that.

**Comment by JDGrimes** [comment]
## External events

An interesting thing to note from https://github.com/WordPoints/hooks-api/issues/63#issuecomment-167581633:

> Another thing to consider is that we'd have to just trust whatever GUID was given us by the other install anyway. We wouldn't be able to verify it. What that basically means is that it would make it very easy for the user to cheat the repeat detection system by supplying made up GUIDs.
> 
> In other words, GUIDs would only fully protect us from repeat fires within the context of the install; added event validation measures would need to be taken for any action that is triggered from the outside. Put another way, all external events are repeatable by default, since we don't provide the infrastructure to detect and stop repetitions.

Basically the point is that events which originate from outside the install (like publishing a post on another site) may not have their given context verified. Some may be verified, but context verification may not really be entirely reliable, and we may not provide any framework for it. Because of this, events accepted from outside of the install can be easily gamed by manipulating the unverified context provided by the outside source when the event occurs. This doesn't exactly mean that our repetition detection API won't work for them; it will, but without verification of the context it is possible to game the system. 

I'm not sure we need to take any action on this information, other than when writing up documentation, but perhaps we should consider the possibility of adding extra roadblocks that must be overcome when creating a non-repeatable external event. That'd likely require us to detect external events somehow though, and I'm not sure how we'd do that.

I suppose the crux of the matter is that any data from the outside should be considered stateful data, unless the source of the data is considered to be fully trusted. That way events which involve data from untrusted sources would always be repeatable.
